Unify validation and submission generation

sevanspowell opened the pull request:
- Split the cardano-metadata-submitter executable into two
  sub-programs. Users can now pass the:
  - "entry" argument (i.e. "cardano-metadata-submitter entry ...") to
    recover the original behaviour.
  - "validate" argument to validate created entries.
- Added the "metadata-server" project as a dependency. It now offers a
  "metadata-lib" library which we use here for it's validation
  component. The validation component provides us with definitions for
  the default rules applicable to all metadata, and means to create
  our own. We define our own wallet-specific validation rules and
  append them to the defaults.
- Created a new "Cardano.Metadata.Validation.Wallet" module to hold
  this new validation logic.
- Add simple Makefile to make entering a development environment
  easier on Nix, plus target to apply styling.
- Move command line parsing and program configuration into 'Config'
  module to keep some separation of responsiblity.

sevanspowell committed:
Unify validation and submission generation

- Split the cardano-metadata-submitter executable into two
  sub-programs. Users can now pass the:
  - "entry" argument (i.e. "cardano-metadata-submitter entry ...") to
    recover the original behaviour.
  - "validate" argument to validate created entries.
- Added the "metadata-server" project as a dependency. It now offers a
  "metadata-lib" library which we use here for it's validation
  component. The validation component provides us with definitions for
  the default rules applicable to all metadata, and means to create
  our own. We define our own wallet-specific validation rules and
  append them to the defaults.
- Created a new "Cardano.Metadata.Validation.Wallet" module to hold
  this new validation logic.
- Add simple Makefile to make entering a development environment
  easier on Nix, plus target to apply styling.
- Move command line parsing and program configuration into 'Config'
  module to keep some separation of responsiblity.

sevanspowell commented:
Draft PR atm to just get it into CI.

gitmachtl reviewed (commented):
request includes again the unit parameter, this is gone, isn't it?

gitmachtl commented on README.md:
```diff
@@ -66,7 +66,7 @@ cardano-metadata-submitter 19309eb9c066253cede617dc635223ace320ae0bbdd5bd1968439
+cardano-metadata-submitter entry 19309eb9c066253cede617dc635223ace320ae0bbdd5bd1968439cd0 \
   --ticker "GIL" \
   --url "https://finalfantasy.fandom.com/wiki/Gil" \
   --unit "2,cents" \
```
unit it gone i guess... ?

sevanspowell replied:
> request includes again the unit parameter, this is gone, isn't it?

Wow great catch. I knew I had to be careful of this but completely forgot to check. Thanks

KtorZ commented on app/Config.hs:
```diff
@@ -0,0 +1,231 @@
+  = MainnetTip
+  | TestnetTip
+  | CustomSlot SlotNo
+  deriving (Eq, Show)
```
I am afraid that is now obsolete ? We used to need the chain tip when validating time slots, yet this has been removed to favor useful use-cases with auto-expiring assets.

sevanspowell replied:
Yes, this can be removed :+1:

KtorZ commented on src/Cardano/Metadata/Validation/Wallet.hs:
```diff
@@ -0,0 +1,91 @@
+{- | Wallet-specific validation rules.
+-}
+
+module Cardano.Metadata.Validation.Wallet
```
I don't get this module's name? There's nothing specific about _the wallet_ per se. If this refers to well-known properties, then we should refer to them as "well known properties" :no_mouth: That the wallet relies on well-known properties is anecdotal, many tools will eventually rely on well-known properties too.

sevanspowell replied:
You have noticed that I have not re-written the existing wallet
validation logic in the style of the validation library provided by
the "metadata-lib" project. If I was to do so, we may well find parts
of the logic that can be shared with other projects, just like some of
the rules I'm already sharing from the "metadata-lib" project (see
the contents of "defaultRules"). I haven't re-written it for a number of reasons
though:
  1. It works
  2. I have limited time
  3. There aren't enough tests around it for me to be confident my
     changes won't introduce an error.

I think I've already done a great deal to introduce a framework around
which we can share rules in the "metadata-lib" project. Further work
can certainly generalise and share some of the wallet's validation
logic, now there is a framework for doing so.

So yeah, you're correct, this logic is not necessarily
wallet-specific. There is some logic in your work that can be shared
with other projects. But from my point of view "validateEntry" is a
black-box which is used to validate wallet entries and cannot be used
to validate non-wallet entries (for example, a metadata entry without
a policy would not pass "validateEntry", and policies are very
specific to the minting of assets).

KtorZ commented on src/Cardano/Metadata/Validation/Wallet.hs:
```diff
@@ -0,0 +1,91 @@
+    mMeta =
+      case diff of
+        Added newFile     -> Just (fileContents newFile)
+        Removed _         -> Nothing
```
I thought removing from the registry was forbidden ?

sevanspowell replied:
The composition of rules should satisfy our requirements, each
individual rule need only satisfy part of our requirements.

Removing from the registry is forbidden, but that's not the concern of
this rule. For example, the "metadata-lib" project provides a
"isJSONFile" rule here
(https://github.com/input-output-hk/metadata-server/blob/aebd96c623aaf32426e9b7189d60728fd3d6b768/metadata-lib/src/Cardano/Metadata/Validation/Rules.hs#L111).
That rule is only concerned with checking that each file has the
".json" extension. From the rule's point of view, it's worth
considering the question, "does this rule care?". In the case of
"isJSONFile", it wants to check that new files have the ".json"
extension, not check that you're only deleting ".json" files, that's a
very different rule. So it consciously decides to allow removed files,
knowing that it's not it's job. 

By keeping each rule "singular", composition of rules is frictionless.
Testing is also simplified.

In the proposed system, the GitHub validator checks that there is no
removal of files. There are already murmurs of allowing removal of
files under some trust scheme, so I think it best to keep that removal
logic in one place.

KtorZ commented on app/Config.hs:
```diff
@@ -0,0 +1,231 @@
+                 , TestnetTip <$ P.string "testnet"
+                 , CustomSlot <$> P.decimal
+                 , pure MainnetTip
+                 ]
```
I think this is now obsolete.

KtorZ reviewed (commented):
There's also a set of automated tests (node.js) to be updated and, make sure that they still pass.
Seems like this PR would actually introduce _another validation logic_ from the metadata-server, but still use the current submitter logic when creating entry? The unification isn't quite clear to me :sweat_smile: ?

gitmachtl commented:
Do the cardano-metadata-submitter need an online connection? Creating the json must be doable in offline mode. Or only required for the validation?

KtorZ commented:
@gitmachtl everything is done offline (creation and validation).

KtorZ commented:
It's called "submitter" but "submitting" is actually the only thing it doesn't do. :| 
It is merely a command-line for constructing, signing and validating token metadata. The submission happens manually via a PR on the metadata registry itself.

gitmachtl commented:
> @gitmachtl everything is done offline (creation and validation).

thx, good. the comments about the tip confused me.

gitmachtl commented:
> It's called "submitter" but "submitting" is actually the only thing it doesn't do. :|
> It is merely a command-line for constructing, signing and validating token metadata. The submission happens manually via a PR on the metadata registry itself.

yep. thx. using it for a while now. operator scripts are also ready for mainnet release. have to add the "entry" parameter in there once its updated.

gitmachtl commented:
> It's called "submitter" but "submitting" is actually the only thing it doesn't do. :|
> It is merely a command-line for constructing, signing and validating token metadata. The submission happens manually via a PR on the metadata registry itself.

cardano-metadata-creator :smiley:

sevanspowell commented:
> There's also a set of automated tests (node.js) to be updated and, make sure that they still pass.
> Seems like this PR would actually introduce _another validation logic_ from the metadata-server, but still use the current submitter logic when creating entry? The unification isn't quite clear to me sweat_smile ?

Roger, we definitely need to get those tests added to the CI.

It's true that the "metadata-lib" project still provides validation
logic in the form of a library, and so the validation logic isn't
unified from a compile-time/dependency point of view. From an
execution point of view though, the "cardano-metadata-submitter" is
completely in charge of which validation rules it uses from the core
library and how it uses them. The validation library is there purely
as support and the "cardano-metadata-submitter" can certainly choose
to ignore it, or even re-implement it, but I figured it's best to
isolate and share what we can with other potential projects when it's
easy to do so.

sevanspowell commented:
> > It's called "submitter" but "submitting" is actually the only thing it doesn't do. :|
> > It is merely a command-line for constructing, signing and validating token metadata. The submission happens manually via a PR on the metadata registry itself.
> 
> cardano-metadata-creator or token/asset-metadata-creator smiley

I really would like to rename both projects.

"metadata-server" implies that it's serving metadata files, but it should probably be named "offchain-metadata-core", both to disambiguate offchain vs. onchain metadata, and to imply that it's a core set of libraries and tools, shareable between projects, which is what I see it as.

The "metadata-server" itself is so un-opinionated about that format of the metadata that I believe it would belong in a core project such as "offchain-metadata-core".

"token/asset-metadata-creator" for "cardano-metadata-submitter" also sounds great.

The only thing stopping me is that a number of end-users know these projects by their current names, and importantly, the "cardano-foundation" has already written a bunch of documentation using these names.

sevanspowell commented:
Just FYI, this pull-request https://github.com/input-output-hk/metadata-server/pull/51 is required for this work, and is the source for many of my comments about a "validation framework" provided by "metadata-lib".

sevanspowell committed:
Fix automated NodeJS tests

sevanspowell committed:
Remove dead slot num code

sevanspowell commented:
I have fixed the automated NodeJS tests and removed the dead slot number code.

As for getting the NodeJS tests into CI, I'm talking with devops about the best way to get that into Hydra. I know there are a number of ways of running NodeJS projects in Nix, but am unsure about the trade-offs of each.

rvl reviewed (approved):
Looks good. Just run the nodejs tests locally and see if they pass.
You don't want to be using Hydra for npm stuff.
